### ISSUE TITLE
Add a way to discover users who are replicating too many docs

### ISSUE THREAD

**Issue by garethbowen** [medic/cht-core#6251]
**What feature do you want to improve?**
In 3.8.0 we added a feature to [warn when replicating too many docs](#5362). This warning is shown when creating a new user, when that user logs in for the first time, and then shown in the API logs every hour the user attempts to replicate.

This is useful but unless people are monitoring the logs closely is likely to be missed.

**Describe the improvement you'd like**
1. Extend the [new monitoring endpoint](#6133) to include the number of users replicating too many docs.
2. Provide a method of finding which users are replicating too many docs so they can be investigated. This could be a new API or an addition to the monitoring API JSON format response.

The easiest way to implement this is to extend the existing logging code so instead of refreshing the cache each hour it retains the logins for longer (perhaps 7 days). This may require persistence (eg: in a `_local` couch doc) so that it's maintained between restarts, particularly if the heavy user is crashing API.

An alternative implementation would be to query the telemetry docs which include the number of docs locally. This data will always be a little out of date (telemetry is synced monthly) and only works for users who have successfully synced at least once so wouldn't include a user whose replication is constantly failing.

Once the monitoring code is in place it may no longer be necessary to log this information.

Once the endpoint is done consider including a warning in the users page of the admin app, or raising an issue to tackle this in a later release.

**Describe alternatives you've considered**
We could provide a script to query couch and find this information, but it's quite a heavy operation so to run it for all users could be expensive. By recording the information when the user logs in it's essentially free.

**Comment by latin-panda** [comment]
This ticket has 2 PRs:

medic-conf: https://github.com/medic/medic-conf/pull/340
cht-docs: https://github.com/medic/cht-docs/pull/234/files

**Comment by latin-panda** [comment]
This ticket has 2 PR:
https://github.com/medic/cht-core/pull/6550
https://github.com/medic/cht-docs/pull/242

**Comment by garethbowen** [comment]
Ready for AT in `6251-logs-replicating-many-docs`.

@latin-panda Can you please add steps for how to AT this issue?

**Comment by latin-panda** [comment]
On this ticket we added a new DB in CouchDB called `medic-logs`, where we add/update a document per user when they replicate docs, this document is called `replication-count-[user name]`.
The `replication-count-[user name]` contains the user name, count of replicated docs and logged date.

To test this I usually do the following requests:
1. Replicate documents with an user that has `offline` role: `GET:  /medic/_changes`
2. Get the full list of logs: `GET:  /api/v1/users-doc-count` here we should see a log with the user from 1st step. 
This endpoint has an optional query parameter, if you want to get logs of 1 user: `GET:  /api/v1/users-doc-count?user=medic` 
3.  If the count exceeds the replication limit (10.000 aprox), then when requesting to the monitor endpoint, it should return the count of users that have exceeds the limit. `GET:  /api/v1/monitoring`

The `replication-count-[user name]` will update if there is a difference between the old and the new on the following:
Log date: 1 month difference
or 
Count: 100 difference

**Comment by ngaruko** [comment]
LGTM
Tested on the branch by replicating 20k + docs to a user and querying the end point `GET: /api/v1/users-doc-count`:
```
{
    "limit": 10000,
    "users": [
        {
            "_id": "replication-count-andra",
            "_rev": "44-bed64752f74832d95ac1d86084f7e326",
            "user": "andra",
            "date": 1600127700641,
            "count": 23139
        },
        {
            "_id": "replication-count-bede_ngaruko",
            "_rev": "1-beec3c1435b940e8675da43e7bf94434",
            "user": "bede_ngaruko",
            "date": 1600053875344,
            "count": 41
        }
    ]
}
``` 
and `/api/v1/users-doc-count?user=andra` returns :
```
{
    "limit": 10000,
    "users": {
        "_id": "replication-count-andra",
        "_rev": "45-4ed6a4695608f4ad7d7fec237b93de7b",
        "user": "andra",
        "date": 1600128174030,
        "count": 27339
    }
}
```
Then checking the `/api/v1/monitoring`** end point return the number of users that exceeded the limit, in this case, only one (not sure if we should return not just the number but also the name - that sounds intuitively more useful @latin-panda ?)
```
curl http://admin:pass@localhost:5988/api/v1/monitoring 
{"version":{"app":"6251-logs-replicating-many-docs","node":"v13.7.0","couchdb":"2.3.1"},"couchdb":{"medic":{"name":"medic","update_sequence":23300,"doc_count":22795,"doc_del_count":7
..
...
"feedback":{"count":501},"conflict":{"count":-1},**"replication_limit":{"count":1}**
```

**Comment by latin-panda** [comment]
Hi @ngaruko, I though about that too, but looks like the intention for monitoring endpoint is to give indicators to the user, then later they can decide if they want to see more detailed information by using`/api/v1/users-doc-count`.
Maybe we can think of an enriching mechanism for the monitoring in another ticket ?
Is this right @garethbowen?

**Comment by garethbowen** [comment]
That's right. The monitoring API is designed to be useful for automated monitoring so we can set up some robot to record the count over time and alert someone if the count goes above some number. This is much harder to configure if we return an array of usernames.

**Comment by ngaruko** [comment]
Thanks @lalkunwar  and @garethbowen. Let's move on 👍
